### PR TITLE
feat: market service rate limiter

### LINKED ISSUE #367: Implement rate limiting in each market service adapter
### Overview

Currently, all our market data service providers have some sort of rate limiting, and we don't have anything in our request logic that respects this. E.g. CoinGecko only supports 50 req/min, and will `429` after than for a minute until you're unblocked.

### Reference

* Here's an example of this happening https://github.com/shapeshift/web/issues/930
* Investigate each of CoinGecko, CoinCap, and Yearn market services docs for their respective rate limits
* Implement rate limiting according to the free tier plans according to each providers docs
* Look at doing this with the axios config/initialization, rather than some handrolled queue mechanism
* Investigate the use of [this axios-rate-limit package](https://www.npmjs.com/package/axios-rate-limit)

### Acceptance Criteria

* The market service will not hit an upstream API data provider at a rate that will exceed the rate limit
* Include unit tests in your implementation demonstrating this, with a dummy number of requests far exceeding the limits
* Link to the providers docs in your PR to substantiate the configured rate limits chosen
* If docs aren't provided, empirically determine what this is by hammering their API until you are rate limited or not

### Ownership

- [X] If my bounty needs engineering or needs product I have added the respective labels on the right
- [X] As the sponsor of this bounty I will review the changes in a preview environment (ops/product) or review the PR (engineering)

### Bounty Hunters

- [Join our Discord](https://discord.gg/shapeshift)
- Include an expected timeline for you to complete work in the work plan when you apply for this bounty!
- Please refer to [this link](https://shapeshiftdao.gitbook.io/getting-started/how-to-contribute/bounties#resources-for-new-contributors-and-bounty-hunters) for some basic info

### PR DESCRIPTION
**Pull request by stackedq**
Rate limiter implemented for market-service package + unit tests
closes #367

### CODE REVIEW

File: packages/market-service/src/utils/rateLimiters.test.ts [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+      const axiosAdapterLimiter = new ServerRateLimiter({
+        tokensPerInterval: axiosAdapterLimiterRate,
+        interval: axiosAdapterLimiterInterval,
+        fireImmediately: true
+      })
+
+      async function axiosTestAdapter(config: AxiosRequestConfig) {
+        const remainingRequests = await axiosAdapterLimiter.removeTokens(1)
+
+        // exeeded rate limiter
```
Comment (gomesalexandre) [inline packages/market-service/src/utils/rateLimiters.test.ts:26]:
```suggestion
        // exceeded rateLimiter limit
```

File: packages/market-service/src/utils/rateLimiters.test.ts [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+      const functionLimiter = new ServerRateLimiter({
+        tokensPerInterval: rateLimiterRate,
+        interval: rateLimiterInterval,
+        fireImmediately: true
+      })
+
+      const functionToBeCalled = async () => {
+        const remainingCalls = await functionLimiter.removeTokens(1)
+        // exeeded rate limiter
```
Comment (gomesalexandre) [inline packages/market-service/src/utils/rateLimiters.test.ts:165]: ditto
```suggestion
        // exceeded rateLimiter limit
```

**Review by gomesalexandre — APPROVED** [review]
Thank you for the contribution @stackedq, again that's really nice work!
A few small comments and the PR title to fix according to [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) but those are nitpicks, the logic looks great so I won't request changes for it.

Props on the full test coverage of the `rateLimiters` module, as it should be. We're testing that the requests are being limited according to the rate limit, but are not trying to actually test it by making requests to the current 3 external services we need to limit requests for. 

Also props on finding the `limiter` npm module. That's exactly what I had in mind when I was thinking of a "poor man's rate-limiting for tests without spinning up an actual server".

Perfect! 👏🏽 👑

File: packages/market-service/package.json [hunk 3]
```diff
@@ -28,8 +28,12 @@
     "@ethersproject/providers": "^5.5.1",
     "@yfi/sdk": "^1.0.27",
     "axios": "^0.26.0",
+    "axios-rate-limit": "^1.3.0",
```
[the remark this answers is not in the document]
Comment (stackedq) [reply packages/market-service/package.json:31]: > better not to reinvent the wheel whenever we can
that's exactly what I had in mind.

[the remark this answers is not in the document]
Comment (stackedq) [reply packages/market-service/package.json:31]: So can I resolve this conversation?

File: yarn.lock [hunk 4]
```diff
@@ -2682,41 +2682,20 @@
   resolved "https://registry.yarnpkg.com/@sindresorhus/is/-/is-0.14.0.tgz#9fb3a3cf3132328151f353de4632e01e52102bea"
   integrity sha512-9NET910DNaIPngYnLLPeg+Ogzqsi9uM4mSboU5y6p8S5DzMTVEsJZrawi+BoDNUVBa2DhJqQYUFvMDfgU062LQ==
 
-"@sinonjs/commons@^1.6.0", "@sinonjs/commons@^1.7.0", "@sinonjs/commons@^1.8.3":
+"@sinonjs/commons@^1.7.0":
```
Comment (gomesalexandre) [inline yarn.lock:2685]: Can you checkout `origin/develop` on this one and rerun the `yarn add` / `yarn install` commands? I tried it locally and I don't get new `yarn.lock` entries

Comment (stackedq) [reply yarn.lock:2685]: sure

Comment (stackedq) [reply yarn.lock:2685]: origin/develop or origin/main?

Comment (gomesalexandre) [reply yarn.lock:2685]: `origin/main` since you have it as destination. You're faster than the snipe edit :D

Comment (stackedq) [reply yarn.lock:2685]: You're right that shouldn't be changed, commits on their way!

**Review by gomesalexandre — APPROVED** [review]

**Review by theoboldfrazier — APPROVED** [review]
Looks good. Ran locally, it works and tests pass

**Review by 0xdef1cafe — APPROVED** [review]
solid, nice work

**Pull request merged**